internal/rbd/snap_diff.go

```diff
@@ -0,0 +1,214 @@
+/*
+Copyright 2025 The Ceph-CSI Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package rbd
+
+import (
+	"context"
+	"errors"
+	"fmt"
+
+	rbderrors "github.com/ceph/ceph-csi/internal/rbd/errors"
+	"github.com/ceph/ceph-csi/internal/rbd/types"
+	"github.com/ceph/ceph-csi/internal/util/log"
+
+	librbd "github.com/ceph/go-ceph/rbd"
+	"github.com/container-storage-interface/spec/lib/go/csi"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+)
+
+const (
+	// TODO: Make internalMaxResults configurable via a flag or config option.
+	// internalMaxResults is the maximum number of results to return in a single
+	// response. This is used to limit the number of changed blocks that needs
+	// to be stored in memory before sending a response to the stream.
+	internalMaxResults = 20
+)
+
+// sendStreamResponse sends the response to the stream based on its type.
+// It handles both GetMetadataAllocated and GetMetadataDelta responses.
+// It returns an error if the stream type is invalid or if sending the response fails.
+func sendStreamResponse(
+	ctx context.Context,
+	stream grpc.ServerStream,
+	volSize int64,
+	changedBlocks []*csi.BlockMetadata,
+) error {
+	if len(changedBlocks) == 0 {
+		return nil
+	}
+
+	log.DebugLog(ctx, "Sending %d changed blocks to the stream", len(changedBlocks))
+	switch s := stream.(type) {
+	case csi.SnapshotMetadata_GetMetadataAllocatedServer:
+		return s.Send(&csi.GetMetadataAllocatedResponse{
+			BlockMetadataType:   csi.BlockMetadataType_VARIABLE_LENGTH,
+			VolumeCapacityBytes: volSize,
+			BlockMetadata:       changedBlocks,
+		})
+	case csi.SnapshotMetadata_GetMetadataDeltaServer:
+		return s.Send(&csi.GetMetadataDeltaResponse{
+			BlockMetadataType:   csi.BlockMetadataType_VARIABLE_LENGTH,
+			VolumeCapacityBytes: volSize,
+			BlockMetadata:       changedBlocks,
+		})
+	default:
+		return errors.New("invalid stream type")
+	}
+}
+
+// StreamMetadata streams the block metadata for a snapshot.
+// If baseRBDSnap is provided, it calculates the delta between the
+// current snapshot and the base snapshot.
+func (rbdSnap *rbdSnapshot) StreamMetadata(
```

Comment:
I have the feeling this function does too much. The `rbdSnapshot` struct should probably not send anything, but instead return `[]*csi.BlockMetadata` to the caller which can then send it out? Would that not make it a cleaner implementation?

Reply:
> I have the feeling this function does too much. 


Broke down this function to include multiple helpers

>The `rbdSnapshot` struct should probably not send anything, but instead return `[]*csi.BlockMetadata` to the caller which can then send it out? Would that not make it a cleaner implementation?

It's a streaming API, so we would be sending more than once.
Used this as inspiration to instead send down a func that only takes `[]*csi.BlockMetadata` as input and send stream response internally.